src/NetworkResolver/NetworkResolverInterface.php

```diff
@@ -0,0 +1,30 @@
+<?php
+
+
+namespace Yiisoft\Yii\Web\NetworkResolver;
+
+use Nyholm\Psr7\Uri;
+use Psr\Http\Message\ServerRequestInterface;
+
+interface NetworkResolverInterface
+{
+
+    /**
+     * @return static
+     */
+    public function withServerRequest(ServerRequestInterface $serverRequest);
```

Comment:
I still think that it's no correct. Server request is not optional. It is required for all interface methods. So, since resolver is meant to be called with different request instances, this method and `getServerRequest()` should be removed and `$request` should be added to `getRemoveIp()`, `getUserIp()`, `isSecureConnection()`.

Reply:
Current `getServerRequest()` code could be moved to middleware.

Reply:
> I still think that it's not correct. Server request is not optional.

With the PSR structure of yii3 I don't see a better solution.

> and $request should be added to getRemoveIp(), getUserIp(), isSecureConnection().

This solution is not developers friendly. And besides that, not really an OOP approach, but rather static functions.

```php
//yii2
$request->getUserIp()
//symfony: (https://symfony.com/doc/current/create_framework/http_foundation.html)
$request->getClientIp()
// symfony: (https://stackoverflow.com/questions/9029757/how-do-i-get-the-user-ip-address-in-symfony2-controller)
$this->container->get('request')->getClientIp();
public function indexAction(Request $request)
{
    $ip = $request->getClientIp();
}

//and yii3:
$container->get('networkResolver')->getUserIp($request);
```

> Current getServerRequest() code could be moved to middleware.

Modifying IPs and request is a one merged process. Separation would mean duplicate code. It might be possible to outsource the common code section, but that would only complicate the code. Can be solved with "BasicNetworkResolver", but not with the more complex IP process.

Reply:
> Current getServerRequest() code could be moved to middleware.

I can imagine splitting it into 2 interfaces and `NetworkResolverInterface` would implement it. And middleware is generalized, eg. `RequestModifier`. There are processes where eg. deleting or modifying headers.

Reply:
Plan B:
The network resolver middleware will have multiple instances and store the IP addresses as an attribute.
```php
// in network resolver middleware
return $request->withAttribute('clientIp', $clientIp)->withAttribute('remoteIp', $remoteIp);
```

@samdark reviews?

Reply:
> //yii3 with more developer friendly
> $container->get('networkResolver')->getUserIp()

Where from would it get server request? It's not in the container.

> UPD

Is there any situation where we need both user IP and server IP at the same time? If not, I'd prefer `NetworkResolver` to modify request directly by changing values without adding any extra attributes.

Reply:
> Is there any situation where we need both user IP and server IP at the same time?

I think there is a case where both are needed. Of course, this is a special use. I think that is perfectly acceptable. And if you need both in one project, you can do it with your own middleware development.

Only very uncomfortable with use.

```php
$ip = $request->getServerParams()['REMOTE_ADDR'];
```

Reply:
> Is there any situation where we need both user IP and server IP at the same time? If not, I'd prefer NetworkResolver to modify request directly by changing values without adding any extra attributes.

I start implementing it based on that and then it becomes middleware only.

Reply:
`BasicNetworkResolver` refactored to middleware is complete.

Reply:
Now I see that the `$server` parameters cannot be overwrite without a new instance (`withServerParams ()` is not defined). Regardless, the middleware implementation is correct, but it is unavoidable to use `$request->withAttributes(...)`. At the framework level, this also means that the attribute name of user's IP should be unified.